rfcs/0008-over-window.md

```diff
@@ -0,0 +1,126 @@
+---
+feature: over_window
+authors:
+  - "TennyZhuang"
+start_date: "2022/10/27"
+---
+
+# Over Window
+
+## Summary
+
+We will introduce the Over Window (OverWindow) to RisingWave in the RFC.
+
+Another well-known name of OverWindow is [Window Function][window_function_wiki]. Unfortunately, the name is highly ambiguous in streaming.
+
+Flink named the feature Over Aggregation][over_aggregation_flink], but we think Over Window can describe the part more concisely.
+
+[window_function_wiki]: https://en.wikipedia.org/wiki/Window_function_(SQL)#:~:text=In%20SQL%2C%20a%20window%20function,single%20value%20for%20multiple%20rows.
+
+## Definitions
+
+Window function call:
+
+```plain
+function_name ([expression [, expression ... ]]) [ FILTER ( WHERE filter_clause ) ] OVER window_name
+function_name ([expression [, expression ... ]]) [ FILTER ( WHERE filter_clause ) ] OVER ( window_definition )
+function_name ( * ) [ FILTER ( WHERE filter_clause ) ] OVER window_name
+function_name ( * ) [ FILTER ( WHERE filter_clause ) ] OVER ( window_definition )
+```
+
+Window definition:
+
+```plain
+[ existing_window_name ]
+[ PARTITION BY expression [, ...] ]
+[ ORDER BY expression [ ASC | DESC | USING operator ] [ NULLS { FIRST | LAST } ] [, ...] ]
+[ frame_clause ]
+```
+
+frame_cause:
+
+```plain
+ RANGE | ROWS | GROUPS } frame_start [ frame_exclusion ]
+{ RANGE | ROWS | GROUPS } BETWEEN frame_start AND frame_end [ frame_exclusion ]
+```
+
+We can simplify that in our first version:
+
+```plain
+[ existing_window_name ]
+[ PARTITION BY expression ]
+[ ORDER BY column_ref ]
+[ frame_clause ]
+```
+
+```plain
+ROWS BETWEEN frame_start AND frame_end
+```
+
+## Motivation
+
+The feature is a part of the SQL standard and is widely supported by different SQL databases and streaming systems.
+
+The feature is also helpful in fraud detection.
+
+## Design
+
+### Primary key derivation
+
+The output rows count of OverWindow is exact as the input rows, as is the primary key.
+
+### Distribution key
+
+In most window definitions, one partition key is specified, and the data should be exchanged by the partition key.
+
+### BatchOverWindow
+
+The implementation is trivial, the only thing we need to concern is that the batch aggregator also needs to support `retract`, which is much more similar to the stream aggregator.
```

Comment:
I'm afraid this is not trivial. If supporting `retract` is not enough for holistic functions, such as `count distinct` and `min/max`